compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/NameCollision.hs

```diff
@@ -0,0 +1,165 @@
+fullyResolve :: Name -> FRName
+fullyResolve = FRName . map T.toLower . \case
+    NModule (ModuleName m) ->
+        m
+    NRecordType (ModuleName m) (TypeConName t) ->
+        m ++ t
+    NVariantType (ModuleName m) (TypeConName t) ->
+        m ++ t
+    NEnumType (ModuleName m) (TypeConName t) ->
+        m ++ t
+    NVariantCon (ModuleName m) (TypeConName t) (VariantConName v) ->
+        m ++ t ++ [v]
+    NEnumCon (ModuleName m) (TypeConName t) (VariantConName v) ->
+        m ++ t ++ [v]
+    NField (ModuleName m) (TypeConName t) (FieldName f) ->
+        m ++ t ++ [f]
+    NChoice (ModuleName m) (TypeConName t) (ChoiceName c) ->
+        m ++ t ++ [c]
+
+-- | State of the name collision checker. This is a
+-- map from fully resolved names within a package to their
+-- original names. We update this map as we go along.
+newtype NCState = NCState (M.Map FRName [Name])
+
+-- | Initial name collision checker state.
+initialState :: NCState
+initialState = NCState M.empty
+
+-- | Try to add a name to the NCState. Returns Error only
+-- if the name results in a forbidden name collision.
+addName :: Name -> NCState -> Either Error NCState
+addName name (NCState nameMap) = do
+    let frName = fullyResolve name
+        oldNames = fromMaybe [] (M.lookup frName nameMap)
+        badNames = filter (nameCollisionForbidden name) oldNames
+    if null badNames then do
+        Right . NCState $ M.insert frName (name : oldNames) nameMap
+    else do
+        Left $ EForbiddenNameCollision
+            (displayName name)
+            (map displayName badNames)
+
+checkName :: MonadGamma m => Name -> S.StateT NCState m ()
+checkName name = do
+    oldState <- S.get
+    case addName name oldState of
+        Left err ->
+            throwWithContext err
+        Right !newState ->
+            S.put newState
+
+checkDataType :: MonadGamma m => ModuleName -> DefDataType -> S.StateT NCState m ()
+checkDataType moduleName DefDataType{..} =
+    case dataCons of
+        DataRecord fields -> do
+            checkName (NRecordType moduleName dataTypeCon)
+            forM_ fields $ \(fieldName, _) -> do
+                checkName (NField moduleName dataTypeCon fieldName)
+
+        DataVariant constrs -> do
+            checkName (NVariantType moduleName dataTypeCon)
+            forM_ constrs $ \(vconName, _) -> do
+                checkName (NVariantCon moduleName dataTypeCon vconName)
+
+        DataEnum constrs -> do
+            checkName (NEnumType moduleName dataTypeCon)
+            forM_ constrs $ \vconName -> do
+                checkName (NEnumCon moduleName dataTypeCon vconName)
+
+checkTemplate :: MonadGamma m => ModuleName -> Template -> S.StateT NCState m ()
+checkTemplate moduleName Template{..} = do
+    forM_ tplChoices $ \TemplateChoice{..} ->
+        checkName (NChoice moduleName tplTypeCon chcName)
+
+-- | Check whether a module satisfies the name collision condition.
+checkModule :: MonadGamma m => Module -> m ()
```

Comment:
Does this detect collisions when the modules don't depend on each other? Please convince me it does by adding a test. :smiley:

Reply:
I think for that we need to check the full package, not just a module? This PR only deals with a module check.